src/palimpzest/solver/solver.py

```diff
@@ -71,273 +86,180 @@ def _fileToPDF(candidate: DataRecord):
+                    dr._stats[td.op_id] = stats
+
+                return [dr]
+
+            elif td.query_strategy == QueryStrategy.BONDED:
+                drs, bonded_query_stats, err_msg = runBondedQuery(candidate, td, self._verbose)
+
+                # if bonded query failed, manually set fields to None
+                if err_msg is not None:
+                    print(f"BondedQuery Error: {err_msg}")
+                    dr = DataRecord(td.outputSchema)
+                    for field_name in td.outputSchema.fieldNames():
+                        setattr(dr, field_name, None)
+                    drs = [dr]
+
+                # if profiling, set record's stats for the given op_id
+                if shouldProfile:
+                    stats["bondedQuery"] = bonded_query_stats
+                    for dr in drs:
+                        dr._stats[td.op_id] = stats
+
+                return drs
+
+            elif td.query_strategy == QueryStrategy.BONDED_WITH_FALLBACK:
+                drs, bonded_query_stats, err_msg = runBondedQuery(candidate, td, self._verbose)
+
+                # if bonded query failed, run conventional query
+                if err_msg is not None:
+                    print(f"BondedQuery Error: {err_msg}")
+                    print("Falling back to conventional query")
+                    dr, conventional_query_stats = runConventionalQuery(candidate, td, self._verbose)
+                    drs = [dr]
+
+                    # if profiling, set conventional query stats
                     if shouldProfile:
-                        dr._stats[op_id] = {"fields": stats}
-
-                    return [dr]
-
-                try:
-                    return runBondedQuery()
-                except Exception as e:
-                    try:
-                        return runConventionalQuery()
-                    except Exception as e:
-                        print(f"Error: {e}")
-                        dr = DataRecord(outputSchema)
-                        for field_name in outputSchema.fieldNames():
-                            setattr(dr, field_name, None)
-                        return [dr]
-            return fn
-
-    def _makeFilterFn(self, inputSchema: Schema, filter: Filter, config: Dict[str, Any], model: Model, prompt_strategy: PromptStrategy, op_id: str, shouldProfile=False):
-            # parse inputs
-            doc_schema = str(inputSchema)
-            doc_type = inputSchema.className()
+                        stats["conventionalQuery"] = conventional_query_stats
+
+                # if profiling, set record's stats for the given op_id
+                if shouldProfile:
+                    stats["bondedQuery"] = bonded_query_stats
+                    for dr in drs:
+                        dr._stats[td.op_id] = stats
+
+                return drs
+
+            # TODO
+            elif td.query_strategy == QueryStrategy.CODE_GEN:
+                raise Exception("not implemented yet")
```

Comment:
Should this be flagged for Zui as a place to add his code?

Reply:
Yes, in PR2 I think I've added such a note.